File: pkg/wordle/knowledge_test.go
// Copyright 2022 Mike Helmick
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package wordle

import (
	"errors"
	"testing"
)

// applyAll is a helper for turns that are expected to succeed.
func applyAll(t *testing.T, k *Knowledge, turns ...[2]string) {
	t.Helper()
	for _, turn := range turns {
		if err := k.Apply(turn[0], turn[1]); err != nil {
			t.Fatalf("Apply(%q, %q) returned unexpected error: %v", turn[0], turn[1], err)
		}
	}
}

func TestNewKnowledgeMatchesEverything(t *testing.T) {
	t.Parallel()

	k := NewKnowledge()
	for _, word := range []string{"CRANE", "TOAST", "SASSY", "ABBEY", "ZYMIC"} {
		if !k.Match(word) {
			t.Errorf("Match(%q) = false on fresh Knowledge, want true", word)
		}
	}
}

// TestApplyAccumulatesAcrossTurns is the regression test for the bug that
// prompted this rewrite: green and yellow constraints used to be rebuilt from
// scratch every round, so a letter confirmed in turn one stopped constraining
// anything in turn two.
func TestApplyAccumulatesAcrossTurns(t *testing.T) {
	t.Parallel()

	k := NewKnowledge()
	// Turn one establishes A in the middle. Turn two says nothing about A at
	// all -- under the old reset-per-round behaviour that was enough to forget
	// it.
	applyAll(t, k,
		[2]string{"CRANE", "..G.."},
		[2]string{"MILKY", "....."},
	)

	if !k.Match("TOAST") {
		t.Error("Match(\"TOAST\") = false, want true: it satisfies both turns")
	}
	if k.Match("ABOUT") {
		t.Error("Match(\"ABOUT\") = true, want false: turn one proved position 3 is A")
	}
	if k.Match("TOADY") {
		t.Error("Match(\"TOADY\") = true, want false: turn two ruled out Y")
	}
}

// TestApplyLearnsExactLetterCounts covers the deduction the old exclusion-set
// model could not express. A gray tile for a letter that is also green or
// yellow in the same guess is not "this letter is absent" -- it is "there are
// no more of them than you have already seen".
func TestApplyLearnsExactLetterCounts(t *testing.T) {
	t.Parallel()

	t.Run("gray duplicate caps the count", func(t *testing.T) {
		t.Parallel()
		k := NewKnowledge()
		// ERASE against a word with exactly one E, in the last position.
		applyAll(t, k, [2]string{"ERASE", "..Y.G"})

		if !k.Match("ABIDE") {
			t.Error("Match(\"ABIDE\") = false, want true: one E, ends in E, contains A off position 3")
		}
		if k.Match("AGLEE") {
			t.Error("Match(\"AGLEE\") = true, want false: the gray E proves there is exactly one E")
		}
	})

	t.Run("green duplicates raise the minimum", func(t *testing.T) {
		t.Parallel()
		k := NewKnowledge()
		// Two green Ss and a gray S: exactly two Ss, at positions 3 and 4.
		applyAll(t, k, [2]string{"SASSY", "..GGG"})

		if !k.Match("MOSSY") {
			t.Error("Match(\"MOSSY\") = false, want true")
		}
		if k.Match("PESTS") {
			t.Error("Match(\"PESTS\") = true, want false: S is not at positions 3 and 4")
		}
	})

	t.Run("a letter absent everywhere is excluded", func(t *testing.T) {
		t.Parallel()
		k := NewKnowledge()
		applyAll(t, k, [2]string{"CRANE", "....."})

		for _, word := range []string{"TRACK", "NOISE", "CHOMP"} {
			if k.Match(word) {
				t.Errorf("Match(%q) = true, want false: it reuses a letter shown absent", word)
			}
		}
		if !k.Match("SILTY") {
			t.Error("Match(\"SILTY\") = false, want true: it avoids every absent letter")
		}
	})
}

func TestApplyBansYellowPositions(t *testing.T) {
	t.Parallel()

	k := NewKnowledge()
	applyAll(t, k, [2]string{"CRANE", "..Y.."})

	if k.Match("TOAST") {
		t.Error("Match(\"TOAST\") = true, want false: A cannot stay at position 3")
	}
	if !k.Match("ADOPT") {
		t.Error("Match(\"ADOPT\") = false, want true: it has an A somewhere other than position 3")
	}
	if k.Match("SHOUT") {
		t.Error("Match(\"SHOUT\") = true, want false: it has no A at all")
	}
}

func TestGuessedWordsAreNotSuggested(t *testing.T) {
	t.Parallel()

	k := NewKnowledge()
	applyAll(t, k, [2]string{"ABIDE", "GGGGG"})

	if !k.Guessed("ABIDE") {
		t.Error("Guessed(\"ABIDE\") = false, want true")
	}
	if k.Match("ABIDE") {
		t.Error("Match(\"ABIDE\") = true, want false: it has already been played")
	}
}

func TestApplyRejectsContradictions(t *testing.T) {
	t.Parallel()

	cases := []struct {
		name  string
		turns [][2]string
	}{
		{
			name: "two different letters green in one position",
			turns: [][2]string{{"CRANE", "G...."}, {"TOAST", "G...."}},
		},
		{
			name: "a letter is green where it was previously ruled out",
			turns: [][2]string{{"CRANE", ".Y..."}, {"TRACE", ".G..."}},
		},
		{
			name: "a known green is reported misplaced",
			turns: [][2]string{{"CRANE", "G...."}, {"CHOMP", "Y...."}},
		},
		{
			name: "a known green is reported absent",
			turns: [][2]string{{"CRANE", "G...."}, {"CHOMP", "....."}},
		},
		{
			name: "a letter is required and forbidden at once",
			turns: [][2]string{{"CRANE", "..Y.."}, {"ABIDE", "....."}},
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			k := NewKnowledge()
			last := len(tc.turns) - 1
			applyAll(t, k, tc.turns[:last]...)

			before := k.String()
			err := k.Apply(tc.turns[last][0], tc.turns[last][1])
			if !errors.Is(err, ErrContradiction) {
				t.Fatalf("Apply(%q, %q) error = %v, want one wrapping ErrContradiction",
					tc.turns[last][0], tc.turns[last][1], err)
			}
			// A rejected turn must be a no-op, or the caller cannot recover by
			// simply re-entering it.
			if after := k.String(); after != before {
				t.Errorf("knowledge changed despite the error:\n before: %s\n  after: %s", before, after)
			}
		})
	}
}

func TestApplyRejectsInvalidInput(t *testing.T) {
	t.Parallel()

	cases := []struct {
		name    string
		guess   string
		pattern string
		wantErr error
	}{
		{"short guess", "CRAN", ".....", ErrInvalidWord},
		{"long guess", "CRANES", ".....", ErrInvalidWord},
		{"digits in guess", "CR4NE", ".....", ErrInvalidWord},
		{"multi-byte rune in guess", "CRÁN", ".....", ErrInvalidWord},
		{"short pattern", "CRANE", "....", ErrInvalidPattern},
		{"long pattern", "CRANE", "......", ErrInvalidPattern},
		{"unknown pattern character", "CRANE", "..X..", ErrInvalidPattern},
		// These used to be handed to regexp.Compile. A character class is
		// exactly five bytes, and a quantifier is a valid pattern length too,
		// so neither was caught by a length check alone.
		{"regex character class as guess", "[A-Z]", ".....", ErrInvalidWord},
		{"regex quantifier as pattern", "CRANE", ".*.*.", ErrInvalidPattern},
		{"regex anchors as pattern", "CRANE", "^...$", ErrInvalidPattern},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			k := NewKnowledge()
			err := k.Apply(tc.guess, tc.pattern)
			if !errors.Is(err, tc.wantErr) {
				t.Fatalf("Apply(%q, %q) error = %v, want one wrapping %v", tc.guess, tc.pattern, err, tc.wantErr)
			}
			if got := k.String(); got != NewKnowledge().String() {
				t.Errorf("invalid input mutated the knowledge: %s", got)
			}
		})
	}
}

func TestApplyNormalizesInput(t *testing.T) {
	t.Parallel()

	upper := NewKnowledge()
	applyAll(t, upper, [2]string{"CRANE", "..G.."})

	lower := NewKnowledge()
	applyAll(t, lower, [2]string{" crane ", " ..g.. "})

	if got, want := lower.String(), upper.String(); got != want {
		t.Errorf("lowercase input produced different knowledge:\n got: %s\nwant: %s", got, want)
	}
}

func TestMatchRejectsMalformedWords(t *testing.T) {
	t.Parallel()

	k := NewKnowledge()
	for _, word := range []string{"", "CRAN", "CRANES", "cr4ne", "CRÁN"} {
		if k.Match(word) {
			t.Errorf("Match(%q) = true, want false", word)
		}
	}
}

File: web/test/knowledge.test.ts
// Copyright 2022 Mike Helmick
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { describe, expect, it } from 'vitest';

import { WORDS } from '../src/data/words';
import {
  ContradictionError,
  InvalidPatternError,
  InvalidWordError,
} from '../src/solver/errors';
import { feedback } from '../src/solver/feedback';
import { suggest } from '../src/solver/filter';
import { Knowledge } from '../src/solver/knowledge';

function knowing(...turns: Array<[string, string]>): Knowledge {
  const k = new Knowledge();
  for (const [guess, pattern] of turns) k.apply(guess, pattern);
  return k;
}

describe('Knowledge', () => {
  it('matches everything before any turn', () => {
    const k = new Knowledge();
    for (const word of ['CRANE', 'TOAST', 'SASSY', 'ZYMIC']) {
      expect(k.match(word)).toBe(true);
    }
  });

  it('keeps green positions across later turns', () => {
    const k = knowing(['CRANE', '..G..'], ['MILKY', '.....']);
    expect(k.match('TOAST')).toBe(true);
    expect(k.match('ABOUT')).toBe(false); // A is pinned to position 3
    expect(k.match('TOADY')).toBe(false); // Y was ruled out in turn two
  });

  it('caps a letter count when a duplicate comes back gray', () => {
    const k = knowing(['ERASE', '..Y.G']);
    expect(k.match('ABIDE')).toBe(true);
    expect(k.match('AGLEE')).toBe(false); // exactly one E, and AGLEE has two
  });

  it('rules out the position of a misplaced letter', () => {
    const k = knowing(['CRANE', '..Y..']);
    expect(k.match('TOAST')).toBe(false); // A cannot stay in the middle
    expect(k.match('ADOPT')).toBe(true);
    expect(k.match('SHOUT')).toBe(false); // no A at all
  });

  it('never suggests a word already played', () => {
    const k = knowing(['ABIDE', 'GGGGG']);
    expect(k.hasPlayed('ABIDE')).toBe(true);
    expect(k.match('ABIDE')).toBe(false);
  });

  it('normalizes case and surrounding whitespace', () => {
    expect(knowing([' crane ', ' ..g.. ']).describe()).toBe(knowing(['CRANE', '..G..']).describe());
  });

  describe('rejects contradictory feedback', () => {
    const cases: Array<[string, Array<[string, string]>]> = [
      ['two different letters green in one position', [['CRANE', 'G....'], ['TOAST', 'G....']]],
      ['a letter green where it was ruled out', [['CRANE', '.Y...'], ['TRACE', '.G...']]],
      ['a known green reported misplaced', [['CRANE', 'G....'], ['CHOMP', 'Y....']]],
      ['a known green reported absent', [['CRANE', 'G....'], ['CHOMP', '.....']]],
      ['a letter required and forbidden at once', [['CRANE', '..Y..'], ['ABIDE', '.....']]],
    ];

    for (const [name, turns] of cases) {
      it(name, () => {
        const k = knowing(...turns.slice(0, -1));
        const before = k.describe();
        const [guess, pattern] = turns[turns.length - 1]!;

        expect(() => k.apply(guess, pattern)).toThrow(ContradictionError);
        // A rejected turn must be a no-op, or the user cannot simply re-enter it.
        expect(k.describe()).toBe(before);
      });
    }
  });

  describe('rejects malformed input', () => {
    const cases: Array<[string, string, string, unknown]> = [
      ['short guess', 'CRAN', '.....', InvalidWordError],
      ['long guess', 'CRANES', '.....', InvalidWordError],
      ['digits in guess', 'CR4NE', '.....', InvalidWordError],
      ['accented letter', 'CRÁNE', '.....', InvalidWordError],
      ['short pattern', 'CRANE', '....', InvalidPatternError],
      ['unknown pattern character', 'CRANE', '..X..', InvalidPatternError],
      // These were handed to a regular expression compiler by the original Go
      // implementation. Both are exactly five characters, so a length check
      // alone let them through.
      ['regex character class as guess', '[A-Z]', '.....', InvalidWordError],
      ['regex quantifier as pattern', 'CRANE', '.*.*.', InvalidPatternError],
      ['regex anchors as pattern', 'CRANE', '^...$', InvalidPatternError],
    ];

    for (const [name, guess, pattern, expected] of cases) {
      it(name, () => {
        const k = new Knowledge();
        const before = k.describe();
        expect(() => k.apply(guess, pattern)).toThrow(expected as never);
        expect(k.describe()).toBe(before);
      });
    }
  });

  it('clones independently', () => {
    const original = knowing(['CRANE', '..G..']);
    const copy = original.clone();
    copy.apply('MILKY', '.....');

    expect(original.match('TOADY')).toBe(true);
    expect(copy.match('TOADY')).toBe(false);
  });

  it('reset returns to a blank slate', () => {
    const k = knowing(['CRANE', '..G..']);
    k.reset();
    expect(k.describe()).toBe(new Knowledge().describe());
    expect(k.match('ZYMIC')).toBe(true);
  });
});

describe('solving real games', () => {
  // Duplicate letters are where the filtering rules are hardest to get right.
  const answers = ['TOAST', 'SASSY', 'ABBEY', 'LEVEL', 'SPEED', 'MUMMY', 'EERIE', 'CRANE'];

  for (const answer of answers) {
    it(`never filters out ${answer}`, () => {
      const k = new Knowledge();
      let guess = 'CRANE';

      for (let round = 0; round < 8 && guess !== answer; round++) {
        const pattern = feedback(guess, answer);
        k.apply(guess, pattern);

        const possible = suggest(k);
        expect(possible, `after guessing ${guess} (${pattern}); ${k.describe()}`).toContain(answer);
        guess = possible[0]!;
      }
    });
  }

  it('agrees with the CLI transcript in the README', () => {
    const k = knowing(['CRANE', '..G..']);
    expect(suggest(k)).toHaveLength(323);
    k.apply('ABAFT', '..G.G');
    expect(suggest(k)).toHaveLength(15);
    k.apply('GHAST', '..GGG');
    expect(suggest(k)).toEqual(['LOAST', 'PLAST', 'TOAST']);
  });

  it('ships the same word list as the Go package', () => {
    expect(WORDS).toHaveLength(12972);
    expect(WORDS[0]).toBe('AAHED');
    expect(WORDS[WORDS.length - 1]).toBe('ZYMIC');
    expect([...WORDS].sort()).toEqual([...WORDS]);
  });
});
